include/boost/gil/detail/math.hpp

```diff
@@ -28,6 +48,158 @@ inline detail::kernel_2d<T, Allocator> get_identity_kernel()
+                {
+                    flip_kernel_col = kernel2.size() - 1 - kernel2_col;
+                    row_boundary = dummy_row + kernel2.size()/2 - flip_kernel_row;
+                    col_boundary = dummy_col + kernel2.size()/2 - flip_kernel_col;
+                    
+                    // ignore input samples which are out of bound
+                    if (row_boundary >= 0 && row_boundary < convolved_kernel_size &&
+                        col_boundary >= 0 && col_boundary < convolved_kernel_size)
+                    {
+                        aux_total +=
+                            kernel2[flip_kernel_row][flip_kernel_col] *
+                            dummy_kernel[row_boundary][col_boundary];
+                    }
+                }
+            }
+            convolved_kernel[dummy_row][dummy_col] = aux_total;
+        }
+    }
+    return convolved_kernel;
+}
+
+/// \brief Fills kernel vector given as argument with a second order kernel in horizontal or
+/// vertical direction. The type of the kernel which is to be used for filling will be indicated 
+/// by the variable 'type'.
+/// \param kernel - Kernel vector which will be filled.
+/// \param type - Indicates the type of second order derivative kernel which is to be filled inside
+/// first argument.
+static void kernel_vector_fill(std::vector<std::vector<float>>& kernel, kernel_type type)
+{
+    if(type == kernel_type::sobel_dx)
+    {
+        for(std::ptrdiff_t row = 0;row < 5; ++row)
+        {
+            for(std::ptrdiff_t col = 0;col < 5; ++col)
+            {
+                kernel[row][col] = dx_sobel2[5 * row + col];
+            }
+        }
+    }
+    else if(type == kernel_type::sobel_dy)
+    {
+        for(std::ptrdiff_t row =0;row < 5; ++row)
+        {
+            for(std::ptrdiff_t col = 0;col < 5; ++col)
+            {
+                kernel[row][col] = dy_sobel2[5 * row + col];
+            }
+        }
+    }
+}
+
+/// \brief Passes parameters to 'kernel_convolve_impl()' repeatedly until kernel vector of desired
+/// order is obtained.
+/// \param order - Indicates order of derivative whose kernel vector is to be returned.
+/// \param type - Indicates the type of kernel vector which is to be returned.
+static auto kernel_convolve(unsigned int order, kernel_type type) -> std::vector<float>
+{
+    std::vector<float> convolved_kernel_flatten;
+    std::vector<std::vector<float>> convolved_kernel(5, std::vector<float>(5));
+    
+    kernel_vector_fill(convolved_kernel, type);
+
+    std::vector<std::vector<float>>smoothing_dummy = smoothing_kernel;
+
+    // Variable 'smooth_repetition' will store the number of times we need to convolve 
+    // 'smoothing_dummy' with itself. This number when used as a power of 2 in its exponentiation,
+    // will result in a number which is the largest power of 2 smaller than 'order - 2'.
+    unsigned int smooth_repetition = (unsigned int)log2(order - 2);
```

Comment:
- `std::` missing for `log2`
- `static_cast` is preferred to C-style cast
- make the variable `const` unless it's modified